carma_wm/src/CARMAWorldModel.h

```diff
@@ -0,0 +1,150 @@
+#pragma once
+
+/*
+ * Copyright (C) 2019 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#include <carma_wm/WorldModel.h>
+#include <lanelet2_core/primitives/LineString.h>
+#include "IndexedDistanceMap.h"
+
+namespace carma_wm
+{
+/*! \brief Class which implements the WorldModel interface. In addition this class provides write access to the world
+ * model. Write access is achieved through setters for the Map and Route.
+ *
+ *  Proper usage of this class dictates that the Map and Route object be kept in sync
+ *
+ * NOTE: This class relies on some lanelet::routing::RoutingGraph objects these have to be initialized using
+ * TrafficRules which by default are only defined for Germany. The default traffic participant used is a car. Changing
+ * this is a TODO:
+ */
+class CARMAWorldModel : public WorldModel
+{
+public:
+  /**
+   * @brief Constructor
+   *
+   */
+  CARMAWorldModel();
+
+  /**
+   * @brief Destructor as required by interface
+   *
+   */
+  ~CARMAWorldModel();
+
+  /*! \brief Set the current map
+   *
+   *  \param map A shared pointer to the map which will share ownership to this object
+   */
+  void setMap(lanelet::LaneletMapPtr map);
+
+  /*! \brief Set the current route. This route must match the current map for this class to function properly
+   *
+   *  \param route A shared pointer to the route which will share ownership to this object
+   */
+  void setRoute(LaneletRoutePtr route);
+
+  ////
+  // Overrides
+  ////
+  std::pair<TrackPos, TrackPos> routeTrackPos(const lanelet::ConstArea& area) const override;
+
+  TrackPos routeTrackPos(const lanelet::ConstLanelet& lanelet) const override;
+
+  TrackPos routeTrackPos(const lanelet::BasicPoint2d& point) const override;
```

Comment:
Maybe add a note in the doc comment here about the potential for a discontinuous crosstrack

Reply:
Will do